### PR TITLE
Implement array shapes for Preg::match $matches by-ref parameter

### PR DESCRIPTION
**Pull request by Seldaek**

### CODE REVIEW

**Comment by Seldaek** [comment]
@staabm can  you maybe reset your branch off the one on my fork? This targets 2.x which would be preferable as Composer uses that still.

There are two problems:

- phpstan fails to run somehow https://github.com/composer/pcre/actions/runs/9888823477/job/27313571168?pr=25
- tests also fail https://github.com/composer/pcre/actions/runs/9888823483/job/27313570804?pr=25

I think these might just be outdated things causing problems, but I am not familiar enough to fix it so I'd appreciate if you can take a look.

**Comment by staabm** [comment]
I think the errors are related to the phpunit-bridge, see https://github.com/composer/pcre/pull/24/files#r1648824476

can we get rid of the bridge for composer/pcre 2.x?

**Comment by Seldaek** [comment]
It now runs the phpstan tests without the bridge, so I don't think it's that.

image

File: tests/PHPStanTests/nsrt/preg-match.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+function equalMatch(string $s): void
+{
+    if (Preg::match('/Price: /i', $s, $matches) == 1) {
+        assertType('array{string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{string}', $matches);
+}
```
Comment (staabm) [inline tests/PHPStanTests/nsrt/preg-match.php:50]: we need to drop these tests as these cannot be supported atm.

I tried tackling support for it with https://github.com/phpstan/phpstan-src/pull/3185 but it has some open problems and I think we will not finish it in the near future

Comment (Seldaek) [reply tests/PHPStanTests/nsrt/preg-match.php:50]: Ok I can do that myself thanks.

And any clue about the phpstan runtime failure?

> Error: Internal error: Composer\Pcre\PHPStan\PregMatchFlags::getType(): Argument #1 ($flagsArg) must be of type ?Composer\Pcre\PHPStan\Arg, PhpParser\Node\Arg given, called in /home/runner/work/pcre/pcre/src/PHPStan/PregMatchTypeSpecifyingExtension.php on line 62 while analysing file /home/runner/work/pcre/pcre/src/Regex.php

Comment (Seldaek) [reply tests/PHPStanTests/nsrt/preg-match.php:50]: Oh now looking at that again it looks like a missing use statement..

File: tests/PHPStanTests/nsrt/preg-match.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+function identicalMatch(string $s): void
+{
+    if (Preg::match('/Price: /i', $s, $matches) === 1) {
+        assertType('array{string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{string}', $matches);
+}
```
Comment (staabm) [inline tests/PHPStanTests/nsrt/preg-match.php:40]: you also need to comment this test for the same reason

File: src/PHPStan/PregMatchFlags.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            return null;
+        }
+
+        $internalFlagsTypes = [];
+        foreach ($flagsType->getConstantScalarValues() as $constantScalarValue) {
+            if (!is_int($constantScalarValue)) {
+                return null;
+            }
+
+            $internalFlagsTypes[] = $constantScalarValue | PREG_UNMATCHED_AS_NULL;
```
Comment (staabm) [inline src/PHPStan/PregMatchFlags.php:32]:
```suggestion
            $internalFlagsTypes[] = new ConstantIntegerType($constantScalarValue | PREG_UNMATCHED_AS_NULL);
```

**Comment by Seldaek** [comment]
Alright, thanks @staabm !

**Comment by staabm** [comment]
excellent, thank you

**Pull request merged**